File: client-api/src/main/java/net/yakclient/client/api/annotation/MethodInjection.java
package net.yakclient.client.api.annotation;

import net.yakclient.client.api.InjectionPriorities;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import static net.yakclient.client.api.annotation.InjectionDefaults.SELF_REF;
import static net.yakclient.client.api.annotation.InjectionDefaults.SELF_REF_ACCESS;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface MethodInjection {
    public String methodFrom() default SELF_REF;

    public int access() default SELF_REF_ACCESS;

    public String name() default SELF_REF;

    public String descriptor() default SELF_REF;

    public String signature() default SELF_REF;

    public String exceptions() default SELF_REF;

    public int priority() default InjectionPriorities.DEFAULT;
}

File: client-api/src/main/kotlin/net/yakclient/client/api/annotation/MethodInjection.kt
package net.yakclient.client.api.annotation

import net.yakclient.client.api.annotation.processor.InjectionMetadata
import net.yakclient.client.api.annotation.processor.InjectionOption
import net.yakclient.client.api.annotation.processor.InjectionPriorityOption

// Differs from source injection, source injection requires a method with its signature
// to already exist, and just updates the sources. This will inject a whole new method.
@InjectionMetadata("method")
public annotation class MethodInjection(
    @InjectionOption("self")
    public val from: String, // Class this injection is from
    @InjectionOption("to")
    public val to: String, // Class this injection is to
    @InjectionOption("methodFrom")
    public val methodFrom: String, // Method this injection is from

    @InjectionOption("access")
    public val access: Int, // Access of the created method
    @InjectionOption("name")
    public val name: String, // Name of the created method
    @InjectionOption("description")
    public val description: String, // Description of the created method
    @InjectionOption("signature")
    public val signature: String, // signature of the created method
    @InjectionOption("exceptions")
    public val exceptions: String, // Exceptions of the created method, Delimiter split by ','

    @InjectionPriorityOption
    public val priority: Int
)
